### PR TITLE
Adds Runechat Subsystem - Deprecates Use of Timers for Runechat

### PR DESCRIPTION
**Pull request by bobbah**
## About The Pull Request

This replaces all use of timers for Runechat with a dedicated subsystem. This subsystem uses a similar approach to the timer subsystem, with an event queue that is broken up into buckets for each tick and those buckets just being a circular doubly-linked list. The result is zero dependence on timers.

This is a tad experimental, comments welcome.

## Why It's Good For The Game

Moves Runechat into a more controllable and monitor-able state, removes any overhead that may be related to timers. Should reduce memory overhead of this system in particular as this was implemented with no additional datums.

## Changelog
:cl: bobbahbrown
add: Runechat now runs under its own subsystem, this should improve performance and allow for finer control over its performance.
/:cl:

### CODE REVIEW

**Comment by bobbahbrown** [comment]
can a maintainer let me know if they're good with this to be testmerged once that point occurs as I'd like to testmerge on bagil/highpop server in general to test the subsystem under load prior to fullmerge

**Comment by AffectedArc07** [comment]
May want to make a copy of the timer sanity unit test if this SS functions in much of the same way

**Comment by bobbahbrown** [comment]
Anturk and I testmerged on Bagil for round 142611, seemed to run fine 0% contribution to time dilation with 40-50 players. Messages were being shown and cleaned up as expected. Very promising! I want to see how it performs under heavy load with rather high dilation high players or both.

> May want to make a copy of the timer sanity unit test if this SS functions in much of the same way

I excluded the sanity checking and bucket rebuild for the time being as I felt that the checks were unnecessary if all scheduled events were coming from the single source, the chatmessage datum, whereas for timers they are more necessary as they are coming from so many different sources where a coder may schedule negative time and the like

**Comment by bobbahbrown** [comment]
Forgot to round off the scheduled destruction time to ensure it would be executed if it was scheduled with the equivalent of zero wait for timers. This should fix messages persisting for 30-60 seconds (depending on time dilation) after their scheduled destruction.

Also fixed a mistype which could cause some silliness with scheduled destruction time being updated after EOL has fired.

**Comment by Jared-Fogle** [comment]
> I want to see how it performs under heavy load with rather high dilation high players or both.

A round like this happened today and the messages would go away, but after seemingly significant delay, and I believe some earlier than others (so messages at the top of the list would stay, but ones at the bottom might clear).

**Comment by bobbahbrown** [comment]
Yeah @Jared-Fogle that was from the scheduling bug I fixed after that round that you can see above. Shouldn't happen now.

**Comment by Jared-Fogle** [comment]
2020-07-24T07-26-10
Feels new. Some messages are getting cut off like this.

**Comment by bobbahbrown** [comment]
> 2020-07-24T07-26-10
> Feels new. Some messages are getting cut off like this.

Unrelated to this PR, seems to be a bug with runechat / something feeding text into runechat. I'd guess that something (slurring?) is causing excessive whitespace in the message

**Comment by bobbahbrown** [comment]
this is ready for a full merge when a maintainer is happy with it

File: code/datums/chatmessage.dm [hunk 1]
```diff
@@ -152,16 +161,18 @@
...
-	addtimer(CALLBACK(src, .proc/end_of_life), lifespan - CHAT_MESSAGE_EOL_FADE, TIMER_UNIQUE|TIMER_OVERRIDE)
+	enter_subsystem()
 
 /**
-  * Applies final animations to overlay CHAT_MESSAGE_EOL_FADE deciseconds prior to message deletion
+  * Applies final animations to overlay CHAT_MESSAGE_EOL_FADE deciseconds prior to message deletion,
+  * sets time for scheduling deletion and re-enters the runechat SS for qdeling
   */
 /datum/chatmessage/proc/end_of_life(fadetime = CHAT_MESSAGE_EOL_FADE)
+	eol_complete = scheduled_destruction + CHAT_MESSAGE_EOL_FADE
```
Comment (psykzz) [inline code/datums/chatmessage.dm:173]: Any reason you didn't use `fadetime` here?

`eol_complete = scheduled_destruction + fadetime`

Comment (bobbahbrown) [reply code/datums/chatmessage.dm:173]: there is no reason, that should be fadetime you are correct

**Pull request merged**